packages/flutter/lib/src/painting/matrix_utils.dart

```diff
@@ -22,20 +29,20 @@ class MatrixUtils {
     assert(transform != null);
     final Float64List values = transform.storage;
     // Values are stored in column-major order.
-    if (values[0] == 1.0 && // col 1
-        values[1] == 0.0 &&
-        values[2] == 0.0 &&
-        values[3] == 0.0 &&
-        values[4] == 0.0 && // col 2
-        values[5] == 1.0 &&
-        values[6] == 0.0 &&
-        values[7] == 0.0 &&
-        values[8] == 0.0 && // col 3
-        values[9] == 0.0 &&
-        values[10] == 1.0 &&
-        values[11] == 0.0 &&
-        values[14] == 0.0 && // bottom of col 4 (values 12 and 13 are the x and y offsets)
-        values[15] == 1.0) {
+    if (_nearEqual(values[0], 1.0) && // col 1
+        _nearZero(values[1]) &&
+        _nearZero(values[2]) &&
+        _nearZero(values[3]) &&
+        _nearZero(values[4]) && // col 2
+        _nearEqual(values[5], 1.0) &&
+        _nearZero(values[6]) &&
+        _nearZero(values[7]) &&
+        _nearZero(values[8]) && // col 3
+        _nearZero(values[9]) &&
+        _nearEqual(values[10], 1.0) &&
+        _nearZero(values[11]) &&
+        _nearZero(values[14]) && // bottom of col 4 (values 12 and 13 are the x and y offsets)
+        _nearEqual(values[15], 1.0)) {
```

Comment:
I'm curious, does the increased cost of this function still defray the cost of just doing the matrix transform?

I'm very reluctant to add this kind of logic to our framework. It's ok in tests, where we know the answer isn't going to be used by anything further, but in the framework it could easily end up slowly increasing error over time, e.g. if someone tries to do a rotation iteratively in very small increments, but between each change they happen to normalize their matrices as an optimisation, we could end up stopping them from making progress in weird edge cases where they happen to go under our threshold.

Reply:
I didn't notice these methods were being used as an optimization as well as to provide debugging descriptions of Transforms. I'd propose making separate versions of these methods only used by TransformProperty and the Semantics tree code with names that clarify that they are about providing descriptions of transforms not computing exact values. For example, `describeAsTranslation`, `describeAsScale`. By putting describe in the name I expect it is the clear that there isn't a guarantee that the scale of translation is exactly identical to the matrix.

Reply:
Ah yeah, that would be fine.